### PR TITLE
[SYCLomatic] Support SYCLCompat for CUB

### PR DESCRIPTION
**Pull request by yihwang-nv**
Depends on [[SYCLomatic] Add support to SYCLcompat](https://github.com/oneapi-src/SYCLomatic/pull/2122).
Depends on [[SYCLomatic] Using SYCL compat API in library related helper functions](https://github.com/oneapi-src/SYCLomatic/pull/2207).

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h [hunk 1]
```diff
@@ -158,8 +156,8 @@ exclusive_scan(const Item &item, T input, BinaryOperation binary_op,
 /// perform the scan. \returns value of the reduction using binary_op
 template <typename Item, typename T, class BinaryOperation,
           int VALUES_PER_THREAD>
-__dpct_inline__ T
-reduce(Item item, T (&inputs)[VALUES_PER_THREAD], BinaryOperation binary_op) {
+__compat_service_inline__ T reduce(Item item, T (&inputs)[VALUES_PER_THREAD],
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h]: let's use dpct_inline.

Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h]: done

File: clang/lib/DPCT/APINamesTemplateType.inc [hunk 2]
```diff
@@ -58,94 +58,169 @@ TYPE_REWRITE_ENTRY("cub::TransformInputIterator",
                        HeaderType::HT_DPL_Iterator,
                        TYPE_FACTORY(STR("oneapi::dpl::transform_iterator"),
                                     TEMPLATE_ARG(2), TEMPLATE_ARG(1))))
-
+// cub::ConstantInputIterator
 TYPE_REWRITE_ENTRY("cub::ConstantInputIterator",
-                   HEADER_INSERTION_FACTORY(HeaderType::HT_DPCT_DPL_Utils,
-                   TYPE_FACTORY(STR(MapNames::getLibraryHelperNamespace() + "constant_iterator"), TEMPLATE_ARG(0))))
+                   HEADER_INSERTION_FACTORY(
```
Comment (ziranzha) [inline clang/lib/DPCT/APINamesTemplateType.inc]: You don't need to emit a warning when using dpl helper functions.

File: clang/lib/DPCT/APINamesTemplateType.inc [hunk 3]
```diff
@@ -58,94 +58,169 @@ TYPE_REWRITE_ENTRY("cub::TransformInputIterator",
...
+        HeaderType::HT_DPCT_GROUP_Utils,
+        TYPE_CONDITIONAL_FACTORY(
+            UseSYCLCompat(),
+            WARNING_FACTORY(Diagnostics::UNSUPPORT_SYCLCOMPAT, TYPESTR),
+            TYPE_FACTORY(STR(MapNames::getLibraryHelperNamespace() + "group::exchange"),
+                         TEMPLATE_ARG(0), TEMPLATE_ARG(2)))))
+// cub::BlockLoad
+TYPE_REWRITE_ENTRY(
+    "cub::BlockLoad",
+    HEADER_INSERTION_FACTORY(
```
Comment (ziranzha) [inline clang/lib/DPCT/APINamesTemplateType.inc]: Don't need to insert help headers when using syclcompat.

Comment (yihwang-nv) [reply clang/lib/DPCT/APINamesTemplateType.inc]: Done

File: clang/lib/DPCT/APINamesTemplateType.inc (hunk 2 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/APINamesTemplateType.inc]: Done

**Review by timmiesmith — COMMENTED** [review]
Changes to the files in the dpl_extras folder and dpl_utils.hpp look good to me.

**Review by zhiweij1 — APPROVED** [review]

**Review by ziranzha — APPROVED** [review]

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**